postgis/lwgeom_remove_irrelevant_points_for_view.c

```diff
@@ -0,0 +1,327 @@
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright (C) 2024 Sam Peters <gluser1357@gmx.de>
+ *
+ **********************************************************************/
+
+#include "postgres.h"
+#include "funcapi.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/numeric.h"
+#include "access/htup_details.h"
+
+#include "liblwgeom.h"
+#include "liblwgeom_internal.h"
+
+// ===============================================================================
+// Encode location of value related to min and max. Returns
+// - 0x1 for value < min
+// - 0x2 for min <= value <= max
+// - 0x4 for value > max
+// ===============================================================================
+static int encodeToBits(double value, double min, double max) {
+	return value < min ? 0x1 : value <= max ? 0x2 : 0x4;
+}
+
+// ===============================================================================
+// Helper function for polygon and polyline POINTARRAY's.
+// Removes points being irrelevant for rendering the geometry
+// within a view specified by rectangular bounds.
+// ===============================================================================
+static void removePoints(POINTARRAY *points, GBOX *bounds, bool closed) {
+
+	int npoints, minpoints;
+	double xmin, ymin, xmax, ymax;
+
+    int i, next, w;
+    int vx, vy, vx0, vy0, vx1, vy1, vxall, vyall;
+    bool sameX, sameY, insideX, insideY, inside, skip, clear;
+    POINT4D p, p0, p1;  // current, previous, next; point read/write see ptarray_flip_coordinates
+
+	// point number check
+    npoints = points->npoints;
+    minpoints = closed ? 4 : 2; // min points for each polygon ring or linestring
+	if (npoints < minpoints) {
+		// clear if not expected minimum number of points
+	    points->npoints = 0;
+	    return;
+	}
+
+    xmin = bounds->xmin;
+    ymin = bounds->ymin;
+    xmax = bounds->xmax;
+    ymax = bounds->ymax;
+
+	// get previous point [i-1]
+	if (closed) {
+		getPoint4d_p(points, 0, &p);
+		getPoint4d_p(points, npoints - 1, &p0);
+       	if (p.x != p0.x || p.y != p0.y) return; // requirement for polygons: startpoint equals endpoint. Leave untouched of not met.
+		npoints--; // remove double here, and re-add at the end
+		getPoint4d_p(points, npoints - 1, &p0);
+	}
+	else {
+		getPoint4d_p(points, 0, &p0);  // for linestrings reuse start point
+	}
+	vx0 = encodeToBits(p0.x, xmin, xmax);
+	vy0 = encodeToBits(p0.y, ymin, ymax);
+
```

Comment:
The indentation of the code here seems a little haphazard, should be cleaned up.

Reply:
Should be fixed now, thanks.